routing/index_graph_serializer.hpp

```diff
@@ -0,0 +1,411 @@
+    vector<Section> m_sections;
+  };
+
+  class JointIdEncoder final
+  {
+  public:
+    template <class Sink>
+    void Write(Joint::Id originalId, BitWriter<Sink> & writer)
+    {
+      auto const & it = m_convertedIds.find(originalId);
+      if (it != m_convertedIds.end())
+      {
+        Joint::Id const convertedId = it->second;
+        if (convertedId >= m_count)
+        {
+          MYTHROW(RoutingException,
+                  ("Dublicated joint id should be less then count, convertedId =", convertedId,
+                   ", count =", m_count, ", originalId =", originalId));
+        }
+        writer.Write(kRepeatJointIdBit, 1);
+        // m_count - convertedId is less then convertedId in general.
+        // So write this delta to save file space.
+        WriteDelta(writer, m_count - convertedId);
+        return;
+      }
+
+      m_convertedIds[originalId] = m_count;
+      writer.Write(kNewJointIdBit, 1);
+      ++m_count;
+    }
+
+    Joint::Id GetCount() const { return m_count; }
+
+  private:
+    Joint::Id m_count = 0;
+    unordered_map<Joint::Id, Joint::Id> m_convertedIds;
+  };
+
+  class JointIdDecoder final
+  {
+  public:
+    // Joint::Id count is continuois along entire file.
+    // But deserializer skips some sections, therefore we should recover counter at each section
+    // start.
+    JointIdDecoder(Joint::Id startId) : m_count(startId) {}
+
+    template <class Source>
+    Joint::Id Read(BitReader<Source> & reader)
+    {
+      uint8_t const bit = reader.Read(1);
+      if (bit == kRepeatJointIdBit)
+      {
+        uint32_t const delta = ReadDelta(reader);
+        if (delta > m_count)
+          MYTHROW(RoutingException, ("Joint id delta", delta, "> count =", m_count));
+
+        return m_count - delta;
+      }
+
+      return m_count++;
+    }
+
+    Joint::Id GetCount() const { return m_count; }
+
+  private:
+    Joint::Id m_count;
+  };
+
+  // PointsFilter has two goals:
+  //
+  // 1. Deserialization skips some sections.
+  //    Therefore Joint::Id is not a continuous series of number [0, N).
+  //    PointsFilter converts joint ids to continuous series of number [0, N).
+  //
+  // 2. Some joints connects roads from different models.
+  //    E.g. 2 roads joint: all vehicles road + pedestrian road.
+  //    If we loads in car model, pedestrian road is skipped,
+  //    so joint would have only point from all vehicles road.
+  //    PointsFilter filters such redundant joints.
+
+  class PointsFilter final
+  {
+  public:
+    PointsFilter(IndexGraph & graph, Joint::Id numJoints) : m_graph(graph)
+    {
+      m_entries.assign(numJoints, {kEmptyEntry, 0});
+    }
+
+    void Push(Joint::Id jointIdInFile, RoadPoint const & rp);
+    Joint::Id GetCount() const { return m_count; }
+
+  private:
+    static uint32_t constexpr kEmptyEntry = numeric_limits<uint32_t>::max();
+    static uint32_t constexpr kPushedEntry = kEmptyEntry - 1;
+
+    // Joint::Id and point id are stored in same space.
```

Comment:
I dislike the idea to store Joint::Id and point id in the same space - this looks quite hacky. Is it possible to store them in individual spaces?